charters/QueryService/charter.md

```diff
@@ -0,0 +1,71 @@
+
+# Query Service
+
+
+## Description
+The Query Service is an Application Programming Interface (API) for querying Human Cell Atlas (HCA) bundle metadata with different query languages.
+
+## Objectives
+1. Build, operate, and maintain a metadata API query service that allows a variety of queries to be performed against bundle metadata stored in HCA Data Coordination Platform (DCP) Data Storage Service (DSS)
+1. Develop a query language specification and metadata Extract, Transform, Load (ETL) pattern to accommodate intuitive querying
+1. Conduct user experience (UX) research to validate query language usability, understand remaining usability gaps in DCP data searchability and discoverability
+1. Develop a scalable public API for external indexing and other DSS event consumers
+1. Migrate existing ElasticSearch (ES) query subsystem to this public API
+1. Deploy all query systems as stand-alone, modular services
+1. Make DCP index and query systems agnostic to DCP schema evolution
```

Comment:
It's unclear to me how you will be able to deliver on this objective. I see two broad types of metadata changes; syntactic type changes (renaming or moving of fields) and semantic changes (splitting, merging, addition of new required fields etc). These are described in much more detail in the [metadata-schema](https://github.com/HumanCellAtlas/metadata-schema/blob/master/docs/evolution.md) repo. I can see possibilities of coding out of syntactic changes (and the metadata team have proposed several options to enable schema introspection to avoid a "lowest common denominator" approach) but I do not see how it is possible to be agnostic to semantic schema evolution; furthermore, these are likely to be the interesting changes driven by evolving scientific requirements

Reply:
I think your concerns are on target. We will not be able to isolate _everyone_ from changes in the metadata. Clients will still need to deal with changes in the metadata. However, we have some ideas on how to make the _query systems_ robust to these changes.

Reply:
@tburdett appreciate your skepticism here, and it's totally warranted. The intent of our approach here is to build a *general purpose* metadata indexing layer with zero-configuration accommodation of schema changes. This means it should be able to deal with schema heterogeneity without having to reprogram the underlying database or API with custom schemas or migrations whenever upstream metadata schema changes. That does not mean the downstream metadata consumers stop caring about metadata changes; both syntactic and semantic changes will still pass through to them and require updates (separately, the way to deal with those is by versioning the schema generations and filtering metadata by major schema versions).

Reply:
Great! 100% agreed zero config metadata indexing is the goal. I think we're on the same page, but there's probably a wider conversation here that should involve those on the query service/indexing side and the metadata team.

I think at this point I'd like to call out the prospect of automatic migration of older data (using older schemas) to the most recent schema wherever possible. Is this something you guys would envisage as being in scope for the query service, the DSS, the data portal? I don't feel ingest is the correct home for this; it de facto requires retention of old data which is out of scope for ingest